src/charm_state.py

```diff
@@ -102,12 +103,24 @@ def report_stats(self) -> typing.Union[str, bool, None]:
         """
         return self._synapse_config.report_stats
 
+    @property
+    def database_data(self) -> typing.Optional[typing.Dict]:
+        """Return database data from database relation.
+
+        Returns:
+            Dict: Information needed for setting connection with database.
+        """
+        return self._database_data
```

Comment:
Should this be exposed in a more manageable form? E.g., as another dataclass or perhaps a typed dictionary? That way we get the advantages of type hints etc? What about pydantic validation of the database data?

Reply:
Working on that :) thanks